### PR TITLE
[system-probe] Do not drop connections if ebpf closed batch is full

### PR DESCRIPTION
**Pull request by hmahmood**
### What does this PR do?

Sends an individual connection outside of a batch if the ebpf closed conn batch is full. This is the source of quite a few connection drops since the batch size if small and even a relatively small overload on the close path will lead to drops. This is illustrated by tests failing in the CI environment because they cannot find expected connections in the tracer; there is not a large number of connections at any given time on a kitchen instance, yet we regular see test failures due to this reason.

Sending an individual connection in the perf buffer does not seem to add any cpu overhead. The batching continues to save cpu cycles so should still be there; the single connection is an exceptional case.

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/classification/tracer-maps.h [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #ifndef __PROTOCOL_CLASSIFICATION_TRACER_MAPS_H
 #define __PROTOCOL_CLASSIFICATION_TRACER_MAPS_H
 
+#ifdef USM_C
+#error "tracer-maps.h should not be included from usm.c"
+#endif
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/classification/tracer-maps.h:6]: While that sounds reasonable, can you share the reasons?
if that's a problem of "name collision" we can of course change our file name

File: pkg/network/ebpf/c/tracer/events.h [hunk 2]
```diff
@@ -102,13 +102,15 @@ static __always_inline void cleanup_conn(conn_tuple_t *tup, struct sock *sk) {
         return;
     }
 
+    bpf_perf_event_output(ctx, &conn_close_event, cpu, &conn, sizeof(conn));
```
Comment (guyarb) [inline pkg/network/ebpf/c/tracer/events.h:105]: Worth moving the comments in line 107-108 to be above that change, and explaining in the code what are you doing here

**Review by guyarb — COMMENTED** [review]
CRed

**Comment by guyarb** [comment]
Besides, I do see the same error happened in the kitchen tests of that branch https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/267159971

File: pkg/network/ebpf/c/protocols/classification/tracer-maps.h (hunk 1 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/protocols/classification/tracer-maps.h:6]: Its not, usm.o fails to load if the maps are included here.

**Comment by hmahmood** [comment]
> Besides, I do see the same error happened in the kitchen tests of that branch https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/267159971

Hm.. it looks like the connections are there, not sure why the test is saying they are not.

**Review by guyarb — APPROVED** [review]

File: pkg/network/ebpf/c/protocols/classification/tracer-maps.h (hunk 1 above)
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/classification/tracer-maps.h:6]: I would say it's the normal behavior to fail to load if you declare map that are not defined in the loader
why having a special check ?

shall we merge the remaining tracer-maps.h into `pkg/network/ebpf/c/tracer` ?
and s|protocols/classification/tracer-maps.h|protocols/classification/shared-tracer-maps.h|g for `pkg/network/ebpf/c/tracer/stats.h` `pkg/network/ebpf/c/tracer/events.h`

File: pkg/network/tracer/connection/tracer.go [hunk 3]
```diff
@@ -93,8 +93,8 @@ var ConnTracerTelemetry = struct {
 	telemetry.NewGauge(connTracerModuleName, "connections", []string{"ip_proto", "family"}, "Gauge measuring the number of active connections in the EBPF map"),
 	nettelemetry.NewStatCounterWrapper(connTracerModuleName, "tcp_failed_connects", []string{}, "Counter measuring the number of failed TCP connections in the EBPF map"),
 	nettelemetry.NewStatCounterWrapper(connTracerModuleName, "tcp_sent_miscounts", []string{}, "Counter measuring the number of miscounted tcp sends in the EBPF map"),
-	nettelemetry.NewStatCounterWrapper(connTracerModuleName, "missed_tcp_close", []string{}, "Counter measuring the number of missed TCP close events in the EBPF map"),
-	nettelemetry.NewStatCounterWrapper(connTracerModuleName, "missed_udp_close", []string{}, "Counter measuring the number of missed UDP close events in the EBPF map"),
+	nettelemetry.NewStatCounterWrapper(connTracerModuleName, "missed_tcp_close", []string{}, "Counter measuring the number of TCP close events that were not part of a batch"),
```
[the remark this answers is not in the document]
Comment (hmahmood) [reply pkg/network/tracer/connection/tracer.go]: Yes, good catch

**Review by akarpz — APPROVED** [review]

**Pull request merged**